pkg/apis/seedmanagement/types_managedseed.go

```diff
@@ -0,0 +1,157 @@
+	DisableMergingWithParent *bool
+}
+
+// GardenletDeployment specifies certain gardenlet deployment parameters, such as the number of replicas,
+// the image, etc.
+type GardenletDeployment struct {
+	// ReplicaCount is the number of gardenlet replicas. Defaults to 1.
+	ReplicaCount *int32
+	// RevisionHistoryLimit is the number of old gardenlet ReplicaSets to retain to allow rollback. Defaults to 0.
+	RevisionHistoryLimit *int32
+	// ServiceAccountName is the name of the ServiceAccount to use to run gardenlet pods.
+	ServiceAccountName *string
+	// Image is the gardenlet container image.
+	Image *Image
+	// Resources are the compute resources required by the gardenlet container.
+	Resources *corev1.ResourceRequirements
+	// PodLabels are the labels on gardenlet pods.
+	PodLabels map[string]string
+	// PodAnnotations are the annotations on gardenlet pods.
+	PodAnnotations map[string]string
+	// AdditionalVolumes is the list of additional volumes that should be mounted by gardenlet containers.
+	AdditionalVolumes []corev1.Volume
+	// AdditionalVolumeMounts is the list of additional pod volumes to mount into the gardenlet container's filesystem.
+	AdditionalVolumeMounts []corev1.VolumeMount
+	// Env is the list of environment variables to set in the gardenlet container.
+	Env []corev1.EnvVar
+	// VPA specifies whether to enable VPA for gardenlet. Defaults to true.
+	VPA *bool
+}
+
+// Image specifies container image parameters.
+type Image struct {
+	// Repository is the image repository.
+	Repository *string
+	// Tag is the image tag.
+	Tag *string
+	// PullPolicy is the image pull policy. One of Always, Never, IfNotPresent.
+	PullPolicy *corev1.PullPolicy
+}
+
+// GardenConnectionBootstrap describes a mechanism for bootstrapping gardenlet connection to the Garden cluster.
+type GardenConnectionBootstrap string
+
+const (
+	// GardenConnectionBootstrapServiceAccount means that a temporary service account should be used for bootstrapping gardenlet connection to the Garden cluster.
+	GardenConnectionBootstrapServiceAccount GardenConnectionBootstrap = "ServiceAccount"
+	// GardenConnectionBootstrapToken means that a bootstrap token should be used for bootstrapping gardenlet connection to the Garden cluster.
+	GardenConnectionBootstrapToken GardenConnectionBootstrap = "BootstrapToken"
+)
+
+// ManagedSeedStatus is the status of a ManagedSeed.
+type ManagedSeedStatus struct {
+	// Conditions represents the latest available observations of a ManagedSeed's current state.
+	Conditions []gardencore.Condition
+	// ObservedGeneration is the most recent generation observed for this ManagedSeed. It corresponds to the
+	// ManagedSeed's generation, which is updated on mutation by the API Server.
+	ObservedGeneration int64
+}
+
+const (
+	// ManagedSeedValid is a condition type for indicating whether the ManagedSeed is valid.
+	ManagedSeedValid gardencore.ConditionType = "Valid"
```

Comment:
It is set to `False` if the shoot is not found or cannot be registered as seed, e.g. due to missing DNS domain. See the implementation in `managedseed_reconciler.go`.

Reply:
So it's only used for notifying an operator about invalid configuration and normally the condition would stay `true`, right?
Then I would argue, that those cases will be covered by validation and admission (if I don't miss anything).
If during reconciliation errors like `not found` still arise, we should add an event to the `ManagedSeed` about it. I think, this would be the idiomatic way of doing things in a kubernetes controller.

Reply:
If you say this is the idiomatic way, then let's follow it, I don't have a strong opinion. 

To make sure we don't miss something important: Before implementing it like this, I investigated how similar things are done in Gardener and found e.g. https://github.com/gardener/gardener/blob/2039d7e14aded18da06be6cb759514f5e0500bb2/pkg/gardenlet/controller/controllerinstallation/controllerinstallation_control.go#L157. So from this perspective the approach I am following here is already used in Gardener. I also think it's not a bad approach since it allows you to see errors directly on the resource via the conditions.

Reply:
Yeah, we already do it in some places. Our existing codebase doesn't always follow the idiomatic way.

> I also think it's not a bad approach since it allows you to see errors directly on the resource via the conditions.

I agree. Then let's do the following:
- have a condition type like `RegistrationFailure`, which is supposed to be set when something during the registration went wrong (e.g. shoot cannot be found anymore or something that can only be detected during runtime)
- during normal operation the condition isn't set or is removed when it was set previously

This is the way, the deployment controller does it with the `ReplicaFailure` condition type, if I understand it correctly.
WDYT?

Reply:
OK, it's basically a similar approach but it covers more cases. I will adapt it accordingly.

Reply:
On a second thought, we already have a `SeedRegistered` condition, and `RegistrationFailure` is the inverse, which is not very useful - registration failures are already covered by `SeedRegistered` being set to `False`. I think the `ReplicaFailure` is not a good example here - a "replica failure" can occur in an otherwise "available" deployment, while in our case, we can't have both "registration failure" and "seed registered".

How about keeping the condition mostly as it is, but renaming it to properly reflect what is actually being checked, e.g. `ShootCanBeRegistered` or something similar?

Reply:
In fact, there is little reason to check the shoot DNS settings in the reconciler anymore, since the corresponding code that uses them has been moved from the controller to the admission plugin. The only thing that should be checked is that if the shoot exists at all, so then the condition should be simply `ShootExists`.

Reply:
Hmm, yeah. I guess, I would have to check the controller implementation in detail to give a clear answer here, which I haven't done so far.
Let's resolve this for now and I will bring it up again in the new PR, if I still have some concerns.